src/Microsoft.Extensions.Primitives/StringTokenizer.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.Extensions.Primitives
+{
+    /// <summary>
+    /// Tokenizes a <c>string</c> into <see cref="StringSegment"/>s.
+    /// </summary>
+    public struct StringTokenizer :  IEnumerable<StringSegment>
+    {
+        private readonly string _value;
+        private readonly char _separator;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="StringTokenizer"/>.
+        /// </summary>
+        /// <param name="value">The <c>string</c> to tokenize.</param>
+        /// <param name="separator">The character to tokenize by.</param>
+        public StringTokenizer(string value, char separator)
+        {
+            if (value == null)
+            {
+                throw new ArgumentNullException(nameof(value));
+            }
+
+            _value = value;
+            _separator = separator;
+        }
+
+        public Enumerator GetEnumerator() => new Enumerator(ref this);
```

Comment:
The reason for `ref` is that the compiler will insert copies when accessing a readonly struct. It's probably not an important optimization for what we're doing, but it can't hurt.